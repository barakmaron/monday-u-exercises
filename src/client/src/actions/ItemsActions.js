import { ItemsReducers } from '../reducers/items-entities-reducer';

export const GetTasksAction = () => {
    return dispatch => {
        dispatch(ItemsReducers.GetTodos());
    };
}

export const SetCompleteAction = (id) => {
    return dispatch => {
        Promise.resolve(dispatch(ItemsReducers.CompleteTodo(id))).then(() => {
            dispatch(ItemsReducers.GetTodos());
        })
    };
}

export const SetDeleteAction = (id) => {
    return dispatch => {
        Promise.resolve(dispatch(ItemsReducers.DeleteTodo(id))).then(() => {
            dispatch(ItemsReducers.GetTodos());
        });
    };
}

export const EditAction = (id, text) => {
    return dispatch => {
        Promise.resolve(dispatch(ItemsReducers.EditTodo(id, text))).then(() => {
            dispatch(ItemsReducers.GetTodos());
        });
    }
}

export const AddAction = (task) => {
    return dispatch => {
        Promise.resolve(dispatch(ItemsReducers.AddTodo(task))).then(() => {
            dispatch(ItemsReducers.GetTodos());
        });
    }
}

export const ClearAllAction = () => {
    return dispatch => {
        Promise.resolve(dispatch(ItemsReducers.ClearAll())).then(() => {
            dispatch(ItemsReducers.GetTodos());
        });
    }
}

export const SortByNameAction = () => {
    return dispatch => {
       dispatch(ItemsReducers.SortByName());
    }
}