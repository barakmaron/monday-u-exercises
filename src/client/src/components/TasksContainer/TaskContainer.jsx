import React from 'react';
import PropTypes from  'prop-types';
import Task from '../Task/Task';
import style from './taskcontainer.module.css';
import { Tooltip } from 'monday-ui-react-core';

const TaskContainer = ({ tasks, delete_call, complete_call, edit_call }) => {
  return (
    <ul id={style.todo_tasks_container} className={tasks.length ? undefined : style.empty}>
      {tasks.map((task, index) => {
        const content = (task.done && task.status) ? `Was set as done at: ${task.done}` : `This item is not done it...`;
        return(
          <Tooltip 
            key={`task_id${task.id}tooltip`}
            immediateShowDelay={0} 
            content={content}>
            <Task
            key={`task_id${task.id}`}
            number={index + 1}
            item={task}
            delete_call={() => delete_call(task.id)}
            complete_call={() => complete_call(task.id)}
            edit_call={edit_call}></Task>
          </Tooltip>
        );
      })}
    </ul>
  );
};

TaskContainer.prototype = {
    tasks: PropTypes.array,
    delete_call: PropTypes.func,
    complete_call: PropTypes.func,
    edit_call: PropTypes.func
}; 

TaskContainer.defaultProps = {
    tasks: [],
    delete_call: null,
    complete_call: null,
    edit_call: null
};

export default TaskContainer;
