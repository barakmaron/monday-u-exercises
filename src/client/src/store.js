import { configureStore } from "@reduxjs/toolkit";
import thunkMiddleware from "redux-thunk";
import allReducers from "./reducers";
import logger from 'redux-logger';

export const store = configureStore({
  reducer: allReducers,
  middleware: [thunkMiddleware, logger],
  preloadedState: {}
});
