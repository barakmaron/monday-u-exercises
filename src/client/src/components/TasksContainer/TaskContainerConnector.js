import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { getItems } from '../../selectors/items-entities-selectors';
import TaskContainer from './TaskContainer';

const mapStateToProps = (state, ownProps) => {
    const itemsEntities = getItems(state);
    return { itemsEntities };
};

const mapDispatchToProps = (dispatch, ownProps) => {
    return bindActionCreators({ }, dispatch);
};

export default connect(mapStateToProps, mapDispatchToProps)(TaskContainer);