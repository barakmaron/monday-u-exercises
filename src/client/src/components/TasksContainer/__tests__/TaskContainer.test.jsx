import renderer from 'react-test-renderer';
import TaskContainer from '../TaskContainer';
import { render, screen } from "@testing-library/react";
import { Provider } from "react-redux";
import { store } from "../../../store";
import ItemsActions from '../../../actions/ItemsActions';
import TaskContainerConnector from '../TaskContainerConnector';

it('Render snapshot of TaskContainer', () => {
    const tree = renderer.create(<TaskContainer></TaskContainer>).toJSON();
    expect(tree).toMatchSnapshot();
});

const items = [
    {
        id: 56,
        ItemName: "Take dog out for a walk",
        status: false,
    },
    {
        id: 32,
        ItemName: "Do the dishes",
        status: true,
    },
];

describe("ListContainer", () => {
    test("should render both items (one done and one not)", () => {
        const { asFragment } = render(
            <Provider store={store}>
                <TaskContainer itemsEntities={items} />
            </Provider>
        );
        itemsCheckedTestCase(asFragment);
    });
});

describe("ListContainer with mock GetTodos", () => {
    const mock_get_todos = jest.fn(() => items);
    const mock_get_todos_action = jest.spyOn(ItemsActions, 'GetTodos').mockImplementation(mock_get_todos);
    test("should render both items (one done and one not) with mock get tasks", () => {
        ItemsActions.GetTodos();
        expect(mock_get_todos_action).toBeCalled();
    });
});

function itemsCheckedTestCase(asFragment) {
    expect(screen.getByText("Take dog out for a walk")).toBeInTheDocument();
    expect(screen.getByText("Do the dishes")).toBeInTheDocument();
    const checkbox_checked = asFragment().querySelector('input[type="checkbox"][id="32"]');
    expect(checkbox_checked.checked).toEqual(true);
    const checkbox_unchecked = asFragment().querySelector('input[type="checkbox"][id="56"]');
    expect(checkbox_unchecked.checked).toEqual(false);
}