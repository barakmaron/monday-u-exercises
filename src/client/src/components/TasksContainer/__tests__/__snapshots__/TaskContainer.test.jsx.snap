// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render snapshot of TaskContainer 1`] = `
<ul
  className="empty"
  id="todo_tasks_container"
/>
`;
